### PR TITLE
Add NullField

### PR DESCRIPTION
**Pull request by faissalMT**
### Reasons for making this change

The JSON Schema type 'null' is currently validated and inferred as a type but shows an error when trying to render. This corrects the issue such that null fields will render nothing, allowing for things like additional help text or use as a temporary placeholder field.  
  
#1232

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [X] **I'm adding or updating code**
  - [X] I've added and/or updated tests
  - [ ] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
  - [X] I've run `npm run cs-format` on my branch to conform my code to [prettier](https://github.com/prettier/prettier) coding style
* [X] **I'm adding a new feature**
  - [X] I've updated the playground with an example use of the feature

### CODE REVIEW

**Comment by epicfaace** [comment]
@faissalMT , thanks for the PR. In #1232, you mention that null fields should render the title only and not the input box -- but this PR makes null fields render nothing entirely. Was this intentional?

Additionally, I'm wondering how this change plays with the changes done in #1213. If this PR makes it such that "null" fields render nothing, then will our current handling of nullable types be unintuitive?

(#1213 makes a field with type `["string", "null"]` still render the input box for a string. If "null" fields actually render nothing, could one intuitively expect a field with type `["string", "null"]` to actually render a OneOf widget with either an input box or nothing in it?)

Maybe this isn't really a concern though so I'd like to hear your thoughts @faissalMT @warrenseymour

**Comment by warrenseymour** [comment]
@epicfaace I've not taken the time to clone this locally and experiment, but as far as I can tell, from a code perspective, this *shouldn't clash* with what I did in #1213, since that does a reasonably extensive check that the field type is `[<T>, "null"]` or `["null, <T>]` and then simply passes the type on as `<T>`. So I think in the case presented here, `type: "null"` should be handled as @faissalMT intends.

From a UX perspective, however, I agree that explicitly adding a `Null` widget may lead to confusion and would inclined to push this sort of decision onto the consumer of RJSF - if they want to explicit enable a `Null` widget, it should be on them to opt-into by bringing their own Widget, or at least explicitly enabling that option if it is built-in.

Finally, trying to infer/envisage a legitimate use-case for this I looked at a commit that @faissalMT cited in #1232. It *looks* as if he's just trying to render a piece of inert content in the form. Would a `uiSchema` entry using a custom widget be a reasonable way of solving that issue?

**Comment by faissalMT** [comment]
@epicfaace  
>but this PR makes null fields render nothing entirely. Was this intentional?  
  
RJSF itself is taking care of handling the rendering of those things (as per the playground example I added), so there was no need for me to implement it explicitly.  
  
>I agree that explicitly adding a Null widget may lead to confusion and I would be inclined to push this sort of decision onto the consumer of RJSF  
  
I would expect any native JSON schema type to be renderable by RJSF, if the user doesn't like the handling they can always override it. I think this makes the most sense and is significantly better than an error.  
  
>Would a uiSchema entry using a custom widget be a reasonable way of solving that issue?  
  
In that project we have a certain level of preprocessing that generates a uiSchema, however we must provide it with a non-null type at the moment or RJSF will refuse to render our custom widget. This also works well for creating placeholder fields.

**Comment by epicfaace** [comment]
@faissalMT thanks, I didn't realize that it was rendering the title/description as well.

It makes sense how rjsf should have some kind of default widget for null types. However, my primary concern is this: when you make a field such as `helpText` with `type: null`, then you would expect `helpText: null` to be in the formData. However, in this case, there is no key for `helpText` in the formData.

image

What if we make `helpText` required? Then we get stuck with an "is a required property" error, which is only resolved when you actually manually set `helpText: null` in the formData.

image

Perhaps the solution for this is to make sure that a `NullWidget` calls the `onChange` handler with `null`, so that `helpText: null` is actually set in the formData. I'd be ready to merge this PR if we decide it makes sense to do that.

I don't know if, as @warrenseymour noted, this entirely fits the use case to "render a piece of inert content in the form". Do we want to clutter our formData by adding in extra null keys when we just want to add in additional content in the form? Perhaps the real solution fo this _is_ in the uiSchema though.

**Comment by faissalMT** [comment]
@epicfaace  
  
>Perhaps the solution for this is to make sure that a NullWidget calls the onChange handler with null  
  
I like this solution and have committed it. An alternative could also simply be to use a default value, but that seems more tacky to me.  
  
>Do we want to clutter our formData by adding in extra null keys when we just want to add in additional content in the form? Perhaps the real solution fo this is in the uiSchema.  
  
I feel like modifying uiSchema to work with non-existent fields as if they are existing fields could quickly get ugly. The uiSchema may serve to further expand on the details of this content (such as use of a custom widget etc), but items in the uiSchema should always correspond to a real field or I believe some fundamental assumptions this project has made are broken.

File: src/components/fields/NullField.js [hunk 1]
```diff
@@ -1,5 +1,20 @@
-function NullField(props) {
-  return null;
+import { Component } from "react";
+import PropTypes from "prop-types";
+
+class NullField extends Component {
+  componentWillMount() {
```
Comment (epicfaace) [inline src/components/fields/NullField.js]: Is there a reason why you did this in componentWillMount instead of componentDidMount?

Comment (faissalMT) [reply src/components/fields/NullField.js]: I just forgot it was deprecated

**Comment by faissalMT** [comment]
Bump?

File: src/components/fields/NullField.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { Component } from "react";
+import PropTypes from "prop-types";
+
+class NullField extends Component {
+  componentDidMount() {
+    this.props.onChange(null);
```
Comment (epicfaace) [inline src/components/fields/NullField.js]: Perhaps it would be better to call onChange only if `props.value === undefined`?

Comment (faissalMT) [reply src/components/fields/NullField.js]: Agreed, that behaviour might come in handy.

**Comment by epicfaace** [comment]
@faissalMT there's just one thing, thanks for the reminder. I was hesitant to merge this because unlike other widgets, this is the only widget that automatically mutates the formData without any user input (by calling `onChange` on load). I'm not sure if that could potentially cause problems, or if it is okay -- what do you think?

**Comment by faissalMT** [comment]
@epicfaace The system doesn't really have anything that would distinguish between actors and this is a really simple manipulation so I don't think there'll be any issues.

**Comment by epicfaace** [comment]
Okay, thanks. What do you mean by "the system doesn't really have anything that would distinguish between actors"?

**Comment by faissalMT** [comment]
@epicfaace I mean that nothing outside the widgets themselves in RJSF has implemented any assumptions about what kind of actor is entering the data (although they may have been programmed with that assumption), there would be nothing preventing a widget being set by fetched data for example. Our project currently implements a number of automatically filled out fields and it really isn't an issue for individual widgets acting on their own data. RJSF only cares about which widget is entering the data for which form field, and we're not breaking the assumption that a widget will only change the data of its field.

**Comment by epicfaace** [comment]
Sounds good, thanks for the explanation.

**Comment by epicfaace** [comment]
Should `NullField` have proptypes of the other fields, such as in:

https://github.com/mozilla-services/react-jsonschema-form/blob/ab0be5ec42939fc699a60a3779d743ea595ad496/src/components/fields/StringField.js#L60

**Comment by faissalMT** [comment]
@epicfaace I didn't realise we had a central proptypes object for fields, I've changed it now.

**Comment by epicfaace** [comment]
Thanks! Looks good.

**Comment by tomasgvivo** [comment]
When will this changes be released?

**Comment by rmdevos** [comment]
@epicfaace   Will this PR eventually be merged or will the issue with null as default value be resolved in an entirely different way?

**Comment by epicfaace** [comment]
@rmdevos this PR has already been merged -- however, it only addresses fields with `type: null`, not necessarily fields with `default: null`.

**Pull request merged**